rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,178 @@
+
+Example input:
+
+```json
+{ "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Transform config:
+
+```toml
+[transforms.explode]
+type = "explode"
+source = "array!(.events) ?? []" # will be typechecked at compile-time
+```
+
+Ouput:
+
+```json
+{"message": "foo"}
+{"message": "bar"}
+```
+
+Support for iteration as part of [#6031](https://github.com/timberio/vector/issues/6031) will allow for users to do things like map fields onto each element. An example might look something like:
+
+Input:
+
+```json
+{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Transform config (actual mapping syntax TBA):
+
+```toml
+[transforms.explode]
+type = "explode"
+source = "map(array!(.events), |event| event.host = .host) ?? []"
+```
+
+Output:
+
+```json
+{"host": "foobar", "message": "foo"}
+{"host": "foobar", "message": "bar"}
+```
+
+This is similar to the support that the current [`explode` transform PR](https://github.com/timberio/vector/pull/6545) has for merging in top-level fields when creating events from a subfield that has an array.
+
+
+## Doc-level Proposal
+
+To convert an incoming event into multiple events, the `explode` transform can be used. This transform takes a VRL program where it expects the last expression to return an array.
+
+For example, given an input of:
+
+```json
+{ "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+And a transform of:
+
+```toml
+[transforms.explode]
+type = "explode"
+source = "array!(.events) ?? []"
+```
+
+The following events will be output:
+
+```json
+{"message": "foo"}
+{"message": "bar"}
+```
+
+If any elements in the returned array are not an object, they will be turned into a string and set as the `message` key on event.
+
+## Rationale
+
+This enhances Vector with a native transform capable of transforming an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to this proposal.
+
+## Prior Art
+
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the proposed approach.
+
+## Drawbacks
+
+- Adds an additional transform to be aware of.
+- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
+- This idea of the last line of the remap script being the "return value" could be a bit confusing to users in how it differs from `remap` which uses whatever `.` is at the end of the script. However, this is similar to VRL's use in unit tests and conditions for the `filter` and `route` transforms.
+- Ongoing maintenance burden should be minimal.
+
+## Alternatives
```

Comment:
One other alternative could be a config option on the remap transform like `unnest_field = "events"` that works vaguely like SQL's `UNNEST` operator and is applied after the script is run. 

We built remap to reduce the amount that users had to chain together primitive transforms to do what they want, so this might be a little bit more in that spirit than adding a new primitive transform.